src/decision/src/GpsDecision.cpp

```diff
@@ -17,19 +19,128 @@ GpsDecision::GpsDecision(int argc, char **argv, std::string node_name) {
+    std::string compass_topic_name = "/gps_conversion/compass";
     int refresh_rate = 10;
+    gps_subscriber = public_nh.subscribe(gps_filtered_topic_name, refresh_rate, &GpsDecision::gpsCurrentCallBack, this);
     gps_subscriber = public_nh.subscribe(gps_filtered_topic_name, refresh_rate, &GpsDecision::gpsCallBack, this);
-
+    compass_subcriber = public_nh.subscribe(compass_topic_name, refresh_rate, &GpsDecision::compassCallBack, this);
     // Setup Publisher(s)
     std::string twist_topic = public_nh.resolveName("command");
     uint32_t queue_size = 10;
     twist_publisher = nh.advertise<geometry_msgs::Twist>(twist_topic, queue_size);
 
 }
 
-// This is called whenever a new message is received
+double GpsDecision::distance(const geometry_msgs::Point::ConstPtr& relative_gps) {
+    //double current_x = relative_gps->x;
+    double current_x=0.0;
+    double current_y=0.0;
+    double current_z=0.0;
+
+    double x=relative_gps->x;
+    double y=relative_gps->y;
+    double z=relative_gps->z;
+
+    double distance=sqrt(pow((x-current_x),2)+pow((y-current_y),2)+pow((z-current_z),2));
+
+  return distance;
+}
+
+#define TO_DEGREES 180/(M_PI)
+double GpsDecision::desiredAngle(const geometry_msgs::Point relative_gps,
+                                 float current_heading,geometry_msgs::Point currentPoint) {
+
+    double next_x=relative_gps.x;
+    double next_y=relative_gps.y;
+    //calculate the relative x and y
+    double x=(next_x-currentPoint.x);
+    double y=(next_y-currentPoint.y);
+
+    //double dis=distance(relative_gps);
+    double tan=y/x;
+    double angle=(atan(tan));
+    double AngleRelativeNorth;
+    double desiredAngle;
+    AngleRelativeNorth=angle*TO_DEGREES;
+    if(x>=0 && y>=0) { //in the 1st quadrant
+        AngleRelativeNorth=90.0-AngleRelativeNorth;
+    }
+    else if (x<0 && y<0) { //in the 3rd quadrant
+        AngleRelativeNorth=-(90+AngleRelativeNorth);
+    }
+    else if (x>=0 && y<0) { //In the 4th quadrant
+        AngleRelativeNorth=90-AngleRelativeNorth;
+    }
+    else if(x<=0 && y==0) {
+        AngleRelativeNorth=-90;
+    }
+    else if(x==0 && y<0) {
+        AngleRelativeNorth=180;
+    }
+
+    //calcute the angle relative to current heading
+     if(current_heading>AngleRelativeNorth) {
+         if(current_heading-AngleRelativeNorth<=180) {
+             desiredAngle=-(current_heading-AngleRelativeNorth);
+         }
+         else {
+             desiredAngle=360-(current_heading-AngleRelativeNorth);
+         }
+     }
+     else if(current_heading<=AngleRelativeNorth) {
+
+         desiredAngle=AngleRelativeNorth-current_heading;
+         if(desiredAngle>180) {
+             desiredAngle=-(360-desiredAngle);
+         }
+     }
+    if(x==0 && y==0 && current_heading==0) {
+        desiredAngle==0;
+    }
+
+    return desiredAngle;
+}
+
+void  GpsDecision::rotate(double desiredAngle,double angular_velocity) {
```

Comment:
Please either delete this function if it is not used, or provide a comment explaining why it's here.